File: Java/EvalVisitor.java
import java.util.HashMap;
import java.util.Map;

public class EvalVisitor extends LabeledExprBaseVisitor<Float> {
	/** "memory" for our calculator; variable/value pairs go here */
	Map<String, Float> memory = new HashMap<String, Float>();
	
	/** ID '=' expr NEWLINE */
	@Override
	public Float visitAssign(LabeledExprParser.AssignContext ctx){
		String id = ctx.ID().getText();
		if(visit(ctx.expr()) == null){
			return null;
		}
		float  value = visit(ctx.expr());
		memory.put(id, value);
		return value;
	}
	
	/** expr NEWLINE */
	public Float visitPrintExpr(LabeledExprParser.PrintExprContext ctx){
		Float value = visit(ctx.expr());
		if (value != null){
			System.out.println(value);
		}
		return value;
	}
	
	/** FLOAT */
	@Override
	public Float visitFloat(LabeledExprParser.FloatContext ctx){
		return Float.valueOf(ctx.FLOAT().getText());
	}	
	
	/** expr '**' expr (Potencia) */
	@Override
	public Float visitPower(LabeledExprParser.PowerContext ctx) {
		float base = visit(ctx.expr(0));
		float exponente = visit(ctx.expr(1));
		return (float) Math.pow(base, exponente);
	}
	
	/** '|' expr '|' */
	@Override
	public Float visitAbsolute(LabeledExprParser.AbsoluteContext ctx){
		if(ctx.getStart().getText().equals("|") && !ctx.getStop().getText().equals("|")){
			System.out.println("Error: Falta el | de cierre.");
			return null;
		}
		float value = visit(ctx.expr());
		return Math.abs(value);
	}

	/** expr '//' expr */
	@Override
	public Float visitDivent(LabeledExprParser.DiventContext ctx){
		float num = visit(ctx.expr(0));
		float den = visit(ctx.expr(1));
		if(den == 0){
			System.out.println("Error: Division por 0");
			return null;
		}
		else{
			return (float) Math.floor(num / den);
		}
	}
	
	
	/** expr % expr */
	@Override
        public Float visitModulus(LabeledExprParser.ModulusContext ctx){
                float num = visit(ctx.expr(0));
                float den = visit(ctx.expr(1));
                if(den == 0){
                        System.out.println("Error: Division por 0");
                        return null;
                }
                else{
                        return num % den;
                }
        }

	/** ID */
	@Override
	public Float visitId(LabeledExprParser.IdContext ctx){
		String id = ctx.ID().getText();
		if(memory.containsKey(id)) return memory.get(id);
		return null;
	}

	/** expr op=('*'|'/') expr */
	public Float visitMulDiv(LabeledExprParser.MulDivContext ctx){
		if((null==visit(ctx.expr(0))) || (null==visit(ctx.expr(1)))){
			System.out.println("Error: Argumento vacio");
			return null;
		}
		float left = visit(ctx.expr(0));
		float right = visit(ctx.expr(1));
		if(ctx.op.getType() == LabeledExprParser.MUL) return left * right;
		if(right!=0) return left/right;
		else{
			System.out.println("Error: Division por 0");
			return null;
		}
	}
	
	/** expr op=('+' | '-') expr */
	@Override
	public Float visitAddSub(LabeledExprParser.AddSubContext ctx){
		if((null==visit(ctx.expr(0))) || (null==visit(ctx.expr(1)))){
			System.out.println("Error: Argumento vacio");
			return null;
		}
		float left = visit(ctx.expr(0));
		float right = visit(ctx.expr(1));
		if(ctx.op.getType() == LabeledExprParser.ADD) return left + right;
		return left - right;
	}

	/** '(' expr ')' */
        @Override
        public Float visitParens(LabeledExprParser.ParensContext ctx){
                if(ctx.getStart().getText().equals("(") && !ctx.getStop().getText().equals(")")){
                        System.out.println("Error: Falta el ) de cierre.");
                        return null;
                }
                float value = visit(ctx.expr());
                return value;
        }

	/** '-' expr */
	@Override
	public Float visitNeg(LabeledExprParser.NegContext ctx){
		if(null == visit(ctx.expr())){
			System.out.println("Error: Signo - sin argumento");
			return null;
		}
		float value=visit(ctx.expr());
		return -(value);
	}

	/** '+' expr */
	@Override
	public Float visitPos(LabeledExprParser.PosContext ctx){
		if(null==visit(ctx.expr())){
			System.out.println("Error: Signo + sin argumento");
			return null;
		}
		float value = visit(ctx.expr());
		return value;
	}

}

File: Python/EvalVisitor.py
import sys
from antlr4 import *
from LabeledExprLexer import LabeledExprLexer
from LabeledExprParser import LabeledExprParser
from LabeledExprVisitor import LabeledExprVisitor

class EvalVisitor(LabeledExprVisitor):
	def __init__(self):
		self.memory = {} #Diccionario para almacenar variables
	def visitAssign(self,ctx):
		id = ctx.ID().getText()
		value = self.visit(ctx.expr())
		if value is None:
			print('Error: asignacion vacia')
			return None
		self.memory[id] = value
		return value
	def visitPrintExpr(self, ctx):
		value = self.visit(ctx.expr())
		if value is not None:
			print(value)
		return value
	def visitFloat(self,ctx):
		return float(ctx.FLOAT().getText())
	def visitPower(self, ctx):
		base = self.visit(ctx.expr(0))
		exponente = self.visit(ctx.expr(1))
		return base ** exponente
	def visitDivent(self,ctx):
		num = self.visit(ctx.expr(0))
		den = self.visit(ctx.expr(1))
		if den == 0:
			print("Error: Division por cero")
			return None
		else:
		    return num // den
	def visitModulus(self,ctx):
	    num = self.visit(ctx.expr(0))
	    den = self.visit(ctx.expr(1))
	    if den == 0:
	        print("Error: Division por cero")
	        return None
	    else:
	        return num % den
	def visitAbsolute(self,ctx):
	    text = ctx.getText()
	    if not (text.startswith("|") and text.endswith("|")):
	        print("Error: Parentesis no cerrados correctamente")
	        return None
	    value = self.visit(ctx.expr())
	    return abs(value)
	def visitId(self, ctx):
		id = ctx.ID().getText()
		return self.memory.get(id, None)
	def visitMulDiv(self, ctx):
		left = self.visit(ctx.expr(0))
		right = self.visit(ctx.expr(1))
		if ctx.op.type == LabeledExprParser.MUL:
			return left * right
		else:
			if right == 0:
				print("Error: Division por cero")
				return None
			else:
				return left / right
	def visitAddSub(self,ctx):
		left = self.visit(ctx.expr(0))
		right = self.visit(ctx.expr(1))
		if ctx.op.type == LabeledExprParser.ADD:
			return round(left + right, 10)
		return round(left - right, 10)
	def visitParens(self, ctx):
	    text = ctx.getText()
	    if not (text.startswith("(") and text.endswith(")")):
	        print("Error: Parentesis no cerrados correctamente")
	        return None
	    return self.visit(ctx.expr())
	def visitNeg(self,ctx):
		value = self.visit(ctx.expr())
		return -value if value is not None else None
	def visitPos(self,ctx):
	    value = self.visit(ctx.expr())
	    return value if value is not None else None
	def visitSin(self, ctx):
		value = self.visit(ctx.expr())
		return math.sin(value)
	def visitCos(self, ctx):
		value = self.visit(ctx.expr())
		return math.cos(value)
	def visitTan(self, ctx):
		value = self.visit(ctx.expr())
		return math.tan(value)
	def visitLog3(self, ctx):
		value = self.visit(ctx.expr())
		return math.log(value,3
	
